daemon/api.go

```diff
@@ -2352,6 +2360,80 @@ func postCreateUser(c *Command, r *http.Request, user *auth.UserState) Response
 	}, nil)
 }
 
+type postUserRemoveData struct {
+	Username string `json:"username"`
+}
+
+func postRemoveUser(c *Command, r *http.Request, user *auth.UserState) Response {
+	_, uid, _, err := postRemoveUserUcrednetGet(r.RemoteAddr)
+	if err != nil {
+		return BadRequest("cannot get ucrednet uid: %v", err)
+	}
+	if uid != 0 {
+		return BadRequest("cannot use remove-user as non-root")
+	}
+
+	var removeData postUserRemoveData
+
+	decoder := json.NewDecoder(r.Body)
+	if err := decoder.Decode(&removeData); err != nil {
+		return BadRequest("cannot decode remove-user data from request body: %v", err)
+	}
+
+	// verify request
+	st := c.d.overlord.State()
+	st.Lock()
+	users, err := auth.Users(st)
+	st.Unlock()
+	if err != nil {
+		return InternalError("cannot get user count: %s", err)
+	}
+
+	username := removeData.Username
+	if username == "" {
+		return BadRequest("cannot remove user: 'username' field is empty")
+	}
+
+	// check if user can be removed
+	if _, err := userLookup(username); err != nil {
+		return BadRequest("cannot remove user %s: %s", username, err)
+	}
+	id, err := getUserID(username, users)
+	if err != nil {
+		return BadRequest("cannot remove user %s: user is a system user", username)
```

Comment:
How is the error message related to the error from `getUserID`?

Reply:
Yeah, I suspect we somehow want to surface the `err` from getUserID()

Reply:
Ok, this one requires a small explanation: getUserID() returns an error message telling that the user was not found, but it searches only extrausers (from state). There's another test before that one to check if the user doesn't exist at all (including system users). When the first test passes and the second fails, it means that the user exists, but it's a system user that can't be removed. Passing the original error message there would make little sense because the user actually exists, and changing the error message directly in getUserID() isn't appropriate in that context. To make the code easier to understand, perhaps changing the function name, returning a specific error message and adding a comment line would suffice?

Reply:
getUserID is a small helper defined here, it has only one error case (we can discuss whether it should be that way) but given that, I think the error is correct,  it means the user exist on the system but not in our state